Libplanet/Blockchain/BlockChain.cs

```diff
@@ -208,11 +228,49 @@ IEnumerator IEnumerable.GetEnumerator()
 
             foreach (var hashValue in hashValues)
             {
+                AddressStateMap blockStates = Store.GetBlockStates(hashValue);
+                if (blockStates is null)
+                {
+                    if (completeStates)
+                    {
+                        // Calculates and fills the incomplete states
+                        // on the fly.
+                        foreach (Block<T> b in this)
+                        {
+                            if (!(Store.GetBlockStates(b.Hash) is null))
+                            {
+                                continue;
+                            }
```

Comment:
If we skip `SetStates` when the state is already calculated, the order of `StateReference` and `TxNonce` may be mixed.

Reply:
It's assuming that there are already complete `StateReference`s but only block states are absence.

Reply:
See also 8708fd1de8127a2753f77175ca02b5f83662dbb9.